internal/cmd/beta/session_cmd.go

```diff
@@ -0,0 +1,197 @@
+
+All exported variables during the session will be available to the subsequent commands.
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return autoconfig.InvokeForCommand(
+				func(
+					cmdFactory command.Factory,
+					logger *zap.Logger,
+				) error {
+					defer logger.Sync()
+
+					envs, err := executeDefaultShellProgram(
+						cmd.Context(),
+						cmdFactory,
+						cmd.InOrStdin(),
+						cmd.OutOrStdout(),
+						cmd.ErrOrStderr(),
+						nil,
+					)
+					if err != nil {
+						return err
+					}
+
+					// TODO(adamb): currently, the collected env are printed out,
+					// but they could be put in a session.
+					if _, err := cmd.ErrOrStderr().Write([]byte("Collected env during the session:\n")); err != nil {
+						return errors.WithStack(err)
+					}
+
+					for _, env := range envs {
+						_, err := cmd.OutOrStdout().Write([]byte(env + "\n"))
+						if err != nil {
+							return errors.WithStack(err)
+						}
+					}
+
+					return nil
+				},
+			)
+		},
+	}
+
+	cmd.AddCommand(sessionSetupCmd())
+
+	return &cmd
+}
+
+func executeDefaultShellProgram(
+	ctx context.Context,
+	commandFactory command.Factory,
+	stdin io.Reader,
+	stdout io.Writer,
+	stderr io.Writer,
+	additionalEnv []string,
+) ([]string, error) {
+	envCollector, err := command.NewEnvCollectorFactory().Build()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	cfg := &command.ProgramConfig{
+		ProgramName: defaultShell(),
+		Mode:        runnerv2.CommandMode_COMMAND_MODE_CLI,
+		Env: append(
+			[]string{command.CreateEnv(command.EnvCollectorSessionEnvName, "1")},
+			append(envCollector.ExtraEnv(), additionalEnv...)...,
+		),
+	}
+	options := command.CommandOptions{
+		NoShell: true,
+		Stdin:   stdin,
+		Stdout:  stdout,
+		Stderr:  stderr,
+	}
+	program, err := commandFactory.Build(cfg, options)
+	if err != nil {
+		return nil, err
+	}
+
+	err = program.Start(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = program.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	changed, _, err := envCollector.Diff()
+	return changed, err
+}
+
+func defaultShell() string {
+	shell := os.Getenv("SHELL")
+	if shell == "" {
+		shell, _ = exec.LookPath("bash")
+	}
+	if shell == "" {
+		shell = "/bin/sh"
+	}
+	return shell
+}
+
+func sessionSetupCmd() *cobra.Command {
+	var debug bool
+
+	cmd := cobra.Command{
+		Use:    "setup",
+		Hidden: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return autoconfig.InvokeForCommand(
+				func(
+					cmdFactory command.Factory,
+					logger *zap.Logger,
+				) error {
+					defer logger.Sync()
+
+					out := cmd.OutOrStdout()
+
+					if err := requireEnvs(
+						command.EnvCollectorSessionEnvName,
+						command.EnvCollectorSessionPrePathEnvName,
+						command.EnvCollectorSessionPostPathEnvName,
+					); err != nil {
+						logger.Info("session setup is skipped because the environment variable is not set", zap.Error(err))
+						return writeNoopShellCommand(out)
+					}
```

Comment:
The order is that `beta session` starts a shell process with those envs provided first. This shell process starts and executes, in the case of zsh, the `.zshrc` file inside which `beta session setup` is called. 

It's perfectly normal that it typically will error and output no-op shell command in `.zshrc`, but from the user point of view it should be invisible. We can recommend protecting `source <(/path/to/runme beta session setup)` with a conditional checking `command.EnvCollectorSessionEnvName` to avoid that.